File: src/cxx-dev/LiENa/lienaStructure/lienaMessageStructure/lienaMessageQueue.cpp
#include "lienaMessageQueue.h"


/**
 * @brief lienaMessageQueue::lienaMessageQueue
 */
lienaMessageQueue::lienaMessageQueue()
{
    this->deviceID = 0;
    this->messageSequence = new QVector<lienaCustomizedMessage*>();
}

//! -----------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::setDeviceId
//! \param deviceID
//!
void lienaMessageQueue::setDeviceId(uint32_t deviceID){
    this->deviceID = deviceID;
}

//! -----------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::getDeviceId
//! \return
//!
uint32_t lienaMessageQueue::getDeviceId(){
    return this->deviceID;
}

//! ------------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::clear
//!
void lienaMessageQueue::clear(){
    mutex.lock();
    messageSequence->clear();
    mutex.unlock();
}

//! ------------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::isEmpty
//! \return
//!
bool lienaMessageQueue::isEmpty(){
    bool ret = false;
    mutex.lock();
    if(messageSequence->size()==0){
        ret = true;
    }
    mutex.unlock();
    return ret;
}

//! ------------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::append
//! \param msg
//!
void lienaMessageQueue::append(lienaCustomizedMessage*msg){
    mutex.lock();
    messageSequence->append(msg);
    mutex.unlock();
}

//! ------------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::getSize
//! \return
//!
int lienaMessageQueue::getSize(){
    int ret = -1;
    mutex.lock();
    ret = messageSequence->size();
    mutex.unlock();
    return ret;
}

//! ------------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::pop_front
//! \return
//!
lienaCustomizedMessage* lienaMessageQueue::pop_front(){
    lienaCustomizedMessage* ret = nullptr;
    mutex.lock();
    ret = messageSequence->at(0);
    messageSequence->removeFirst();
    mutex.unlock();
    return ret;
}

//! ------------------------------------------------------------------------------------------
//!
//! \brief lienaMessageQueue::pop_back
//! \return
//!
lienaCustomizedMessage* lienaMessageQueue::pop_back(){
    lienaCustomizedMessage* ret = nullptr;
    mutex.lock();
    ret = messageSequence->at(getSize()-1);
    messageSequence->removeLast();
    mutex.unlock();
    return ret;
}

File: src/py-dev/LiENa/LiENaStructure/LiENaMessageStructure/lienaMessageQueue.py
import threading

from RCPContext.LienaControlInstruction import LienaControlInstruction


class LienaMessageQueue:
    def __init__(self):
        self.msgQueueLock = threading.Lock()
        self.msgQueue = list()
        self.index = 0

    def set_device_id(self, device_id):
        self.device_id = device_id

    def get_device_id(self):
        return self.device_id

    def append(self, msg):
        self.msgQueueLock.acquire()
        self.msgQueue.append(msg)
        self.msgQueueLock.release()

    def is_empty(self):
        empty = 0
        if len(self.msgQueue) > 0:
            empty = 1
        return empty

    def get_size(self):
        self.msgQueueLock.acquire()
        length = len(self.msgQueue)
        self.msgQueueLock.release()
        return length

    def pop_front(self):
        ret = None
        self.msgQueueLock.acquire()
        if len(self.msgQueue) > 0:
            ret = self.msgQueue.pop(0)
        self.msgQueueLock.release()
        return ret

    def pop_back(self):
        ret = None
        self.msgQueueLock.acquire()
        if len(self.msgQueue) > 0:
            ret = self.msgQueue.pop(-1)
        self.msgQueueLock.release()
        return ret

    def clear(self):
        self.msgQueueLock.acquire()
        self.msgQueue = []
        self.msgQueueLock.release()


